### DISCUSSION TITLE
Google Authenticator

### DISCUSSION THREAD

**Discussion by jht-prog** [keycloak/keycloak#13795] [Q&A]
Where does keycloak store Google Authenticator credentials?

**Comment by lexcao** [comment]
You can refer this provider
https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java

**Reply by jht-prog** [reply]
thank you.......

**Comment by jht-prog** [comment]
Returning to this topic with a related question, let's assume for now that I'm using the default H2 database for Keycloak. Is there a way to extract user credentials from the database, and in particular, the Google Auth credentials?

**Reply by lexcao** [reply]
Connect your H2 database, and query the following SQL
```
SELECT * FROM CREDENTIAL
WHERE TYPE = 'otp';
```
